sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,304 @@ def transform(self, X):
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Check if % missing in any column > col_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.col_max_missing)):
+            raise ValueError("The following columns have, "
+                             "more than {0}% missing values: {1}"
+                             .format(self.col_max_missing*100, np.where(
+                              mask.sum(axis=0) > (X.shape[0] *
+                                                  self.col_max_missing))))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > col_max_missing
+        bad_rows = mask.sum(axis=1) > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. These "
+                "rows are not included as donor neighbors."
+                    .format(self.row_max_missing*100))
+
+            # Remove rows that have more than row_max_missing % missing
+            X = X[~bad_rows, :]
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
+            raise ValueError("There are only %d samples, "
+                             "but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+
+        # Instantiate NN object, get column means, and store in statistics_
+        neigh = NearestNeighbors(n_neighbors=self.n_neighbors,
+                                 metric=self.metric)
+        self._fitted_neighbors = neigh.fit(X)
+        self.statistics_ = X_col_means
+
+        return self
+
+    def _transform(self, X, n_neighbors_new):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        n_neighbors_new : int
+            Indicates whether to pass n_neighbors or n_neighbors+1 to
+            _tranform().
+            Calling transform() automatically sets this to self.n_neighbors
+            while fit_transform() sets it to self.n_neighbors + 1.
+        """
+        # Import(s) here to avoid circular import
+        from ..neighbors.base import _get_weights
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
```

Comment:
Not covered by tests.

Reply:
line 122 of test_knn_imputation.py or did u mean something else?

Reply:
Well, I mean that if you install https://github.com/codecov/browser-extension or otherwise view the test coverage results, you will see that this line is not covered. Thus if there is a test you *think* is covering this line, there is an exception being raised earlier that is shadowing this line, and your test is only passing because you test error type without testing message.

Reply:
ouch! think i need a doctor for my busted nose...heheh

Reply:
This isn't fixed yet...